### DISCUSSION TITLE
How to unit test Activities?

### DISCUSSION THREAD

**Discussion by wakuflair** [elsa-workflows/elsa-core#244] [Q&A]
I dont' want to use workflow stuff, just test an activity.

I searched the repo, but no test code founded.

**Comment by sfmskywalker** [comment]
You should be able to instantiate your activity and then call its `ExecuteAsync` method, but I do realize it might be less than straightforward to instantiate the `WorkflowExecutionContext`. Writing tests is on my list of things to do, so I will make sure to include proper unit tests for the activities that come with Elsa out of the box.

**Comment by PeterLaComb-Availity** [comment]
I have been thinking about tests for Elsa a lot lately, and I think the value provided by covering the Core of Elsa with test is significant, I will likely put up a PR against this issue in the next few weeks. Hopefully, that would also influence testing of activities in a way.

Is there a preference for test framework? NUnit, XUnit, Other?

**Comment by sfmskywalker** [comment]
100% agree about the value of having tests covering the core. I do recommend doing so for Elsa 2 however, as Elsa 1 is currently in maintenance mode. Not to say that tests are useless for projects under maintenance of course.

**Comment by PeterLaComb-Availity** [comment]
Understood. I'll focus on the 2.0 branch. Perhaps we can back-port some of the result if there is value.

**Comment by sfmskywalker** [comment]
Much appreciated!

**Comment by fulganta** [comment]
Hi,
I'm using Elsa 2.2.1
I succeed to test Elsa activities doing like this:
```
//Call myActivity
var myActivity= new MyActivity(param1, param2, param3);
myActivity.quoteRequestResultInput = quoteRequest;
// Build context
var cancelToken = new CancellationToken(false);
var context = new ActivityExecutionContext(null, null, null, null, false, cancelToken);
// call ExecuteAsync
var res = await myActivity.ExecuteAsync(context);
```
The issue is when I have some activities that need a `context.CorrelationId`. As you said @sfmskywalker , it seems that I need to instantiate  WorkflowExecutionContext that needs a serviceProvider that only have an internal constructor. It makes it quite complex to build.

My approach was to do:
```
//Build WorkflowExecutionContext
var cancelToken = new CancellationToken(false);
var workflowInstance = new WorkflowInstance();
workflowInstance.CorrelationId = "7aca8cd2ac2e44658841b47b3ce2b71a";
IEnumerable<ServiceDescriptor> serviceDescriptors = new List<ServiceDescriptor>();
ServiceProviderOptions options = new ServiceProviderOptions();
ServiceProvider serviceProvider = new ServiceProvider();
var workflowBlueprint = new WorkflowBlueprint();
var workflowExecutionContext = new WorkflowExecutionContext(serviceProvider, workflowBlueprint, workflowInstance);

// Build context
var context = new ActivityExecutionContext(null, workflowExecutionContext, null, null, false, cancelToken);

// call ExecuteAsync
var res = await bookingMurexActivity.ExecuteAsync(context);
```
But I can't instantiate `ServiceProvider`
I'm trying now to do it by reflection but it's not ideal.
Any thought about it?

**Reply by fulganta** [reply]
I succeeded to unit test an activity in case you are interested. 
Mainly relied on test provided in the elsa repo: elsa-core\test\unit\Elsa.UnitTests\Activities\Signaling\SignalReceivedTests.cs
If you download the sources and run the unit test you can find many examples.
Here is how I did an activity unit test. I removed all the personal variables and using field
It's probably not perfect and I didn't finish all, but it can help I believe to tackle unit testing Activities.
```
namespace JadeOrchestrator.tests.ElsaActivities
{
	public class TestActivity
	{
		public TestmyActivity(
			ITestOutputHelper testOutputHelper, Action<IServiceCollection>? configureServices = default)
		{
			TestOutputHelper = testOutputHelper;

			var services = new ServiceCollection().AddElsa(
				options => options.AddConsoleActivities(Console.In, new XunitConsoleForwarder(testOutputHelper)));

			configureServices?.Invoke(services);
			ServiceProvider = services.BuildServiceProvider();
			ServiceScope = ServiceProvider.CreateScope();
			WorkflowBuilder = ServiceScope.ServiceProvider.GetRequiredService<IWorkflowBuilder>();
		}

		protected ITestOutputHelper TestOutputHelper { get; }
		protected ServiceProvider ServiceProvider { get; }
		protected IServiceScope ServiceScope { get; }
		protected IWorkflowBuilder WorkflowBuilder { get; }

		[Fact]
		public void TestBookingMurexSuspendOK()
		{
			//Act
			Task.Run(
					async () =>
					{
						//New myActivity
						var myActivity = new MyActivity();

						// Build WorkflowExecutionContext
						var cancelToken = new CancellationToken(false);
						var workflowInstance = new WorkflowInstance();
						workflowInstance.CorrelationId = "7aca8cd2ac2e44658841b47b3ce2b71a";
						var workflow = new NoOpWorkflow();
						var workflowBlueprint = WorkflowBuilder.Build(workflow);

						var workflowExecutionContext = new WorkflowExecutionContext(
							ServiceScope.ServiceProvider, workflowBlueprint, workflowInstance);

						var context = new ActivityExecutionContext(
							ServiceScope.ServiceProvider, workflowExecutionContext, workflowBlueprint, default, false,
							cancelToken);

						// call ExecuteAsync
						var res = await myActivity.ExecuteAsync(context);
						var resJson = res.ToJson();
						Assert.Contains("Done", resJson);
					})
				.GetAwaiter()
				.GetResult();
		}
	}

	public class NoOpWorkflow : IWorkflow
	{
		public void Build(IWorkflowBuilder builder) { }
	}
}
```
